src/tree.rs

```diff
@@ -214,32 +218,40 @@ fn _bipartition_tree(
             .collect();
 
         if unseen_neighbors.len() == 1 {
-            // This will be false if root
+            // At leaf, will be false at root
+            let pop = pops[node.index()];
+
+            // Update neighbor pop
             let neighbor = unseen_neighbors[0];
             pops[neighbor.index()] += pop;
-            let mut current_partition_tracker = same_partition_tracker[node.index()].clone();
-            same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
 
-            if !node_queue.contains(&neighbor) {
-                node_queue.push_back(neighbor);
+            // Check if balanced; mark as seen
+            if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
+                balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
             }
+            seen_nodes.insert(node.index());
+
+            // Update neighbor partition tracker
+            let mut current_partition_tracker =
+                mem::take(&mut same_partition_tracker[node.index()]);
+            same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+
+            // Queue neighbor
+            node_queue.push_back(neighbor);
         } else if unseen_neighbors.is_empty() {
-            // root
+            // Is root
             break;
         } else {
-            // Not at the leaves of the unseen subgraph
+            // Not a leaf yet
             continue;
         }
-
-        // Check if balanced
-        if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
-            balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
-        }
-
-        seen_nodes.insert(node.index());
     }
 
+    // Convert LinkedList back to vec
     balanced_nodes
+        .iter()
+        .map(|(node, partition_nodes)| (*node, partition_nodes.iter().copied().collect()))
+        .collect()
```

Comment:
This offsets most of the gains you get from using a LinkedList (if any)

Reply:
Yeah, I just perf tested it, and you're right.